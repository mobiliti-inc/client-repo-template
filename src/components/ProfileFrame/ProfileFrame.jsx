import { withRouter } from 'react-router';
import { connect } from 'react-redux';

import { scroll, logOut, refreshAccessToken, uploadUserProfilePhoto } from '../../actions';
import Header from '../Header/Header';
import Footer from '../Footer/Footer';

import styles from './ProfileFrame.scss';
import { LocationModel } from '../../models';
import verifyToken from '../../utils/verifyToken';
import LoadingSpinner from '../../components/LoadingSpinner/LoadingSpinner';
import ProfileWrapper from '../../pages/Profile/common/ProfileWrapper/ProfileWrapper';
import handleAuthRequiredApiCall from '../../utils/handleAuthRequiredApiCall';

const NO_FOOTER_ROUTES = ['/'];

let lastScrollY = 0;

class ProfileFrame extends React.PureComponent {
	static propTypes = {
		children: PropTypes.node.isRequired,
		location: LocationModel.isRequired,
		scroll: PropTypes.func.isRequired,
		scrolling: PropTypes.bool.isRequired,
		isLoggedIn: PropTypes.bool.isRequired,
		logOut: PropTypes.func.isRequired,
		requiresAuth: PropTypes.bool,
		accessToken: PropTypes.string,
		refreshToken: PropTypes.string,
		refreshAccessToken: PropTypes.func.isRequired,
		history: PropTypes.shape({
			push: PropTypes.func.isRequired,
		}).isRequired,
		subscriber: PropTypes.shape({
			profile_photo: PropTypes.string,
		}).isRequired,
		uploadUserProfilePhoto: PropTypes.func.isRequired,
	};

	static defaultProps = {
		requiresAuth: false,
		accessToken: '',
		refreshToken: '',
	}

	componentWillMount = () => {
		this.handleVerifyToken();
	}

	componentDidMount() {
		window.scrollTo({ top: 0, behavior: 'smooth' });
		window.addEventListener('scroll', this.handleScroll);
	}

	componentWillUnmount() {
		window.removeEventListener('scroll', this.handleScroll);
	}

	handleVerifyToken = () => {
		const { accessToken, refreshAccessToken, refreshToken, requiresAuth } = this.props;
		if ((!accessToken || !refreshAccessToken) && requiresAuth) {
			return this.props.history.push('/sign-in');
		}
		if (!verifyToken(accessToken) && requiresAuth) {
			refreshAccessToken(refreshToken, this.props.history);
		}
	}

	handleScroll = () => {
		lastScrollY = window.pageYOffset;
		const scrolling = lastScrollY > 0;
		this.props.scroll(scrolling);
	};

	toggleShowVideo = () => this.setState({ showVideo: !this.state.showVideo })

	handleUploadUserProfilePhoto = (imageData) => {
		const {
			accessToken,
			refreshToken,
			refreshAccessToken,
			uploadUserProfilePhoto,
			history,
		} = this.props;
		handleAuthRequiredApiCall(
			accessToken,
			refreshToken,
			history,
			uploadUserProfilePhoto,
			refreshAccessToken,
			imageData
		);
	}
	render() {
		const {
			location: { pathname },
			isLoggedIn,
			logOut,
			children,
			accessToken,
			requiresAuth,
			scrolling,
			subscriber,
		} = this.props;

		const noFooter = NO_FOOTER_ROUTES.some(el => pathname === el);
		if (!verifyToken(accessToken) && requiresAuth) {
			return (
				<div styleName="loader-container">
					<LoadingSpinner show blue parentClass={styles.loader} />
				</div>);
		}
		return (
			<div styleName="profile-frame">
				<Header
					logOut={logOut}
					isLoggedIn={isLoggedIn}
					scrolling={scrolling}
					styleName={scrolling ? 'profile-header--scrolled' : 'profile-header'}
				/>
				<div styleName="profile-container">
					<main styleName="content">
						<ProfileWrapper
							profilePhoto={subscriber.profile_photo}
							handleUploadUserProfilePhoto={this.handleUploadUserProfilePhoto}
						>
							{children}
						</ProfileWrapper>
					</main>
				</div>
				{
					!noFooter &&
					<Footer
						logOut={logOut}
						isLoggedIn={isLoggedIn}
					/>
				}
			</div>
		);
	}
}

const mapStateToProps = state => ({
	scrolling: state.scrolling,
	isLoggedIn: state.auth.isLoggedIn,
	accessToken: state.auth.userData.accessToken,
	refreshToken: state.auth.userData.refreshToken,
	subscriber: state.subscriber.data.user,
});

export default withRouter(
	connect(
		mapStateToProps,
		{ scroll, logOut, refreshAccessToken, uploadUserProfilePhoto, }
	)(CSSModules(ProfileFrame, styles, { allowMultiple: true }))
);
