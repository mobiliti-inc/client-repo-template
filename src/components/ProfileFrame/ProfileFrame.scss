@import "~variables";

.profile-frame {
	background: rgba(0,0,0,0);
	display: flex;
	flex-direction: column;
}

.profile-container {
	display: flex;
	flex-direction: row;
	flex-grow: 1;
	min-height: 100vh;
	overflow: hidden;
}

.profile-header {
	background: $black;
	position: relative;
	&--scrolled {
		background: rgba(0,0,0,0.5);
	}
}

.content {
	display: flex;
	flex: 1;
	align-items: flex-start;
	flex-wrap: wrap;
	justify-content: space-around;
	width: 100%;
	max-width: 100%;
	background-size: 450px;
}

.loader-container {
	display: flex;
	justify-content: center;
	align-items: center;
	align-content: center;
	height: 100vh;
}

.loader, .loader div {
	width: 60px;
	height: 60px
}
