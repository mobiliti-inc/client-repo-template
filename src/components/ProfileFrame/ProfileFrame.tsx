import { withRouter } from 'react-router';
import { connect } from 'react-redux';
import { scroll, logOut, refreshAccessToken, uploadUserProfilePhoto } from '../../actions';
import Header from '../Header/Header';
import Footer from '../Footer/Footer';
import styles from './ProfileFrame.scss';
import { LocationModel } from '../../models';
import verifyToken from '../../utils/verifyToken';
import LoadingSpinner from '../../components/LoadingSpinner/LoadingSpinner';
import ProfileWrapper from '../../pages/Profile/common/ProfileWrapper/ProfileWrapper';
import handleAuthRequiredApiCall from '../../utils/handleAuthRequiredApiCall';

const NO_FOOTER_ROUTES = ['/'];
let lastScrollY = 0;

type ProfileFrameProps = {
	location: any,
	scroll: (...args: any[]) => any,
	scrolling: boolean,
	isLoggedIn: boolean,
	logOut: (...args: any[]) => any,
	requiresAuth?: boolean,
	accessToken?: string,
	refreshToken?: string,
	refreshAccessToken: (...args: any[]) => any,
	history: {
		push: (...args: any[]) => any
	},
	subscriber: {
		profile_photo?: string
	},
	uploadUserProfilePhoto: (...args: any[]) => any
};

class ProfileFrame extends React.PureComponent<ProfileFrameProps, {}> {
	componentWillMount = () => {
		this.handleVerifyToken();
	};
	componentDidMount() {
		window.scrollTo({ top: 0, behavior: 'smooth' });
		window.addEventListener('scroll', this.handleScroll);
	}
	componentWillUnmount() {
		window.removeEventListener('scroll', this.handleScroll);
	}
	handleVerifyToken = () => {
		const { accessToken, refreshAccessToken, refreshToken, requiresAuth } = this.props;
		if ((!accessToken || !refreshAccessToken) && requiresAuth) {
			return this.props.history.push('/sign-in');
		}
		if (!verifyToken(accessToken) && requiresAuth) {
			refreshAccessToken(refreshToken, this.props.history);
		}
	};
	handleScroll = () => {
		lastScrollY = window.pageYOffset;
		const scrolling = lastScrollY > 0;
		this.props.scroll(scrolling);
	};
	toggleShowVideo = () => this.setState({ showVideo: !this.state.showVideo });
	handleUploadUserProfilePhoto = imageData => {
		const {
			accessToken,
			refreshToken,
			refreshAccessToken,
			uploadUserProfilePhoto,
			history
		} = this.props;
		handleAuthRequiredApiCall(
			accessToken,
			refreshToken,
			history,
			uploadUserProfilePhoto,
			refreshAccessToken,
			imageData
		);
	};
	render() {
		const {
			location: { pathname },
			isLoggedIn,
			logOut,
			children,
			accessToken,
			requiresAuth,
			scrolling,
			subscriber
		} = this.props;
		const noFooter = NO_FOOTER_ROUTES.some(el => pathname === el);
		if (!verifyToken(accessToken) && requiresAuth) {
			return (
				<div styleName="loader-container">
					<LoadingSpinner show blue parentClass={styles.loader} />
				</div>
			);
		}
		return (
			<div styleName="profile-frame">
				<Header
					logOut={logOut}
					isLoggedIn={isLoggedIn}
					scrolling={scrolling}
					styleName={scrolling ? 'profile-header--scrolled' : 'profile-header'}
				/>
				<div styleName="profile-container">
					<main styleName="content">
						<ProfileWrapper
							profilePhoto={subscriber.profile_photo}
							handleUploadUserProfilePhoto={this.handleUploadUserProfilePhoto}
						>
							{children}
						</ProfileWrapper>
					</main>
				</div>
				{!noFooter && <Footer logOut={logOut} isLoggedIn={isLoggedIn} />}
			</div>
		);
	}
}

const mapStateToProps = state => ({
	scrolling: state.scrolling,
	isLoggedIn: state.auth.isLoggedIn,
	accessToken: state.auth.userData.accessToken,
	refreshToken: state.auth.userData.refreshToken,
	subscriber: state.subscriber.data.user
});

export default withRouter(
	connect(mapStateToProps, { scroll, logOut, refreshAccessToken, uploadUserProfilePhoto })(
		CSSModules(ProfileFrame, styles, { allowMultiple: true })
	)
);
