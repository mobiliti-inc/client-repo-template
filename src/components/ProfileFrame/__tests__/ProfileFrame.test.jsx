import ProfileFrame from '../ProfileFrame';


const defaultProps = {
	profilePhoto: 'fake photo',
	uploadUserProfilePhoto: jest.fn(),
};
const component = <ProfileFrame {...defaultProps} />;

describe('ProfileFrame component', () => {
	it('should render', () => {
		const tree = shallow(component);
		matchSnapshot(tree);
	});
});
