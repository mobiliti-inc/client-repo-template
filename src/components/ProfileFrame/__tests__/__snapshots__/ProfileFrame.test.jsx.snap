// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileFrame component should render 1`] = `<Route />`;
